[libunwind][AArch64] Protect PC within libunwind's context.

DanielKristofKiss opened the pull request:
Libunwind manages the registers/context including the program counter which is used effectively as a return address.
__libunwind_Registers_arm64_jumpto can go anywhere where the given buffer 's PC points to.
To increase the robustness of libunwind let's protect the stored address with PAC. Let's use the A key and the base address of the context/registers as modifier. After this patch the PC must be signed therefore the context harder to craft outside of libunwind.

The register set is internal to libunwind and this change is not visible on the APIs.

DanielKristofKiss committed:
[libunwind][AArch64] Protect PC within libunwind's context.

Libunwind manages the regiser context including the program counter
which is used effectively as return address.
To increase the robustness of libunwind let's protect the stored address
with PAC. Since there is no unwind info for this let's use the A key and
the base address of the context/registers as modifier.
__libunwind_Registers_arm64_jumpto can go anywhere where the given buffer
's PC points to. After this patch it needs a signed PC therefore the context
 is more harder to craft outside of libunwind.

The register value is internal to libunwind and the change is not visible
on the the APIs.

DanielKristofKiss commented:
sorry for the debug here, locally I can't reproduce the test failure( macos m1, same toolchain, same build steps )

john-brawn-arm commented on libunwind/src/Registers.hpp:
```diff
@@ -1845,8 +1884,14 @@ class _LIBUNWIND_HIDDEN Registers_arm64 {
+  uint64_t getIP() const { return auth(_registers.__pc, getAuthSalt()); }
+  void setIP(uint64_t value) {
+    // First authenticate the current value of the IP to ensure the context
+    // is still valid. This also ensure the setIP can't be used for signing
```
"ensure the" should be "ensures that"

john-brawn-arm commented on libunwind/src/Registers.hpp:
```diff
@@ -1823,9 +1823,48 @@ extern "C" void *__libunwind_cet_get_jump_target() {
+  }
+#if defined(_LIBUNWIND_IS_NATIVE_ONLY)
+  // Authenticate the given pointer and return with the raw value
+  // if the authentication is succeeded.
```
"authentication is succeeded" should be either "authentication succeeded" or "authentication is successful".

john-brawn-arm commented on libunwind/src/Registers.hpp:
```diff
@@ -1823,9 +1823,48 @@ extern "C" void *__libunwind_cet_get_jump_target() {
 
 class _LIBUNWIND_HIDDEN Registers_arm64 {
+protected:
+  /// The program counter is used effectively as a return address
```
"used effectively" should be "effectively used", or just "used".

john-brawn-arm commented:
I'm not especially familiar with how pointer authentication works with exception unwinding, but this looks OK to me with some minor comments.

asl commented on libunwind/src/Registers.hpp:
```diff
@@ -1845,8 +1884,14 @@ class _LIBUNWIND_HIDDEN Registers_arm64 {
+    // First authenticate the current value of the IP to ensure the context
+    // is still valid. This also ensure the setIP can't be used for signing
+    // arbitrary values.
+    auth(_registers.__pc, getAuthSalt());
```
This potentially introduces a signing gadget to libunwind as you are not checking the result of auth. On systems without FPAC / PAuth2 this is a problem.

DanielKristofKiss replied:
A system without TBI and FPAC could be problematic as [signing](https://developer.arm.com/documentation/ddi0602/2024-09/Shared-Pseudocode/aarch64-functions-pac?lang=en#impl-aarch64.InsertPAC.6) adds PAC to the top bits ( ```result = ((ptr<63:56> EOR PAC<63:56>) : bit55 : ptr<54:0>); ```)
While XPAC's [Strip](https://developer.arm.com/documentation/ddi0602/2024-09/Shared-Pseudocode/aarch64-functions-pac?lang=en )) throws those way. (```original_ptr =  extfield<63:bottom_PAC_bit> : A<bottom_PAC_bit-1:0>;```)

I had a version 88b512720752c31374ca75b4c2ccab8b00b25fb1 for that cross checked the auth and xpac result but it [failed on bots](https://github.com/llvm/llvm-project/actions/runs/11524630174/job/32086008467#step:5:5362) as they don't have FPAC nor TBI and even PAC disabled but there are bits in the top part of the LR. 
xpac drop those bits always regardless of PAC status.

I'm planning to make this configurable, e.g when compiled with `-mbranch-protection`. 

I'm happy to hear more ideas.

DanielKristofKiss committed:
Improve the signing authentication scheme.

Added a descriminator to the buffer.
Signed the second part of the PC with the first part.
misc fix:wq use hint instead of xplaclri

DanielKristofKiss committed:
add always inline attribute to the update_pc.

also move define the descriminator instead of magic value.

asl commented on libunwind/src/Registers.hpp:
```diff
@@ -1845,8 +1986,14 @@ class _LIBUNWIND_HIDDEN Registers_arm64 {
-  void      setIP(uint64_t value) { _registers.__pc = value; }
+  uint64_t getIP() const { return authPC(&_registers, getDiscriminator()); }
+  void setIP(uint64_t value) {
+    // First authenticate the current value of the IP to ensure the context
```
Will you please expand how this would work? As far as I can see, the `value` above comes from outside (e.g. from `setRegister` call). Is this just saved unsigned PC that is located somewhere on the stack? So it could be substituted and then organized in ROP-like chain during unwinding process?

DanielKristofKiss replied:
unwinding starts from a context which is created by the `__unw_getcontext`. 
Here the authentication ensure the context/registers are still pointing to a legitimate address.
without this check any buffer could be used and setIP would just set the arbitrary address and where the restore context would happily jump.
If the `value` is altered outside of libunwind then nothing to do here. Here I assume rest of the application has no idea about PAC. With PAuthAbi this would be different.

ojhunt commented:
I've finally got our downstream hardening to the point where it should be sane: https://github.com/llvm/llvm-project/pull/143230
